로그 저장 시스템 구축

SeoYoonP opened the pull request:
### 📌 PR 타입(하나 이상의 PR 타입을 선택해주세요)
- 기능 추가
- 기능 삭제
- 리팩토링

### 📌 관련 이슈
#31

### ✨ 반영 브랜치
feat/31 -> develop

### ➕ 추가 메모
https://www.notion.so/1ffa3f519ab8805faa34c2d663eebb31?pvs=4 
서비스에 관해 작성한 리포트 입니다.

- LogService같은 경우 책임이 많아 리팩토링을 많이 해야합니다.

### 🐛 테스트 결과 (테스트 결과가 있다면 넣어주세요)
<img width="493" alt="image" src="https://github.com/user-attachments/assets/9126568f-6e3c-4ada-964c-ee0e4235bfe5" />

SeoYoonP committed:
feat: #31 로그데이터 생성 및 수정 서비스 구현

SeoYoonP committed:
feat: #31 jdbc Config 정의

SeoYoonP committed:
feat: #31 Log 생성 및 저장 단위 테스트

SeoYoonP committed:
feat: #31 Log 생성 스케줄러 서비스 정의

SeoYoonP committed:
refactor: #31 스케줄링 어노테이션 추가

SeoYoonP committed:
refactor: #31 로그 생성 서비스 및 테스트코드 삭제

SeoYoonP committed:
feat: #31 로그 데이터 배치 요청을 위한 DTO 정의

SeoYoonP committed:
feat: #31 데이터 수신을 위한 로그 컨트롤러 구현

SeoYoonP committed:
feat: #31 수신 로그 데이터 저장 서비스 구현

SeoYoonP committed:
refactor: #31 @EnableScheduling 어노테이션 제거

SeoYoonP committed:
feat: #31 LogDataSaveServiceTest 구현

- 배치 로그 데이터 저장 성공
- 예외 발생시 로그 데이터 롤백

Jeongyounghyeon commented on src/main/java/org/thisway/log/LogDataSaveService.java:
로그 관리에 방향에서 다음과 같이 메서드화, 상수화 방향으로 의견 드립니다!
```java
@Slf4j
@Service
@RequiredArgsConstructor
public class LogDataSaveService {

    public static final String POWER_LOG_INSERT_SQL = """
        INSERT INTO power_log (
            vehicle_id,
            mdn,
            power_status,
            power_time,
            gps_status,
            latitude,
            longitude,
            "total_trip_meter"
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """;
    public static final String GPS_LOG_INSERT_SQL = """
        INSERT INTO gps_log (
             vehicle_id,
             mdn,
             gps_status,
             latitude,
             longitude,
             anger,
             speed,
             total_trip_meter,
             battery_voltage,
             occurred_time
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """;
    public static final String GEOFENCE_LOG_INSERT_SQL = """
        INSERT INTO geofence_log (
            vehicle_id,
            mdn,
            occured_time,
            geofence_group_id,
            geofence_id,
            event_val,
            gps_status,
            latitude,
            longitude,
            angle
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """;

    private final JdbcTemplate jdbcTemplate;

    @Transactional
    public void saveBatchLogData(LogDataBatchRequest request) {
        Long vehicleId = request.vehicleId();
        Long mdn = request.mdn();
        List<LogDataEntry> entries = request.entries();

        log.info("차량 ID: {}에 대한 로그 데이터 배치 저장 시작 (항목 수: {})", vehicleId, entries.size());

        List<Object[]> powerLogBatch = new ArrayList<>();
        List<Object[]> gpsLogBatch = new ArrayList<>();
        List<Object[]> geofenceLogBatch = new ArrayList<>();

        for (LogDataEntry entry : entries) {
            powerLogBatch.add(toPowerLogParams(vehicleId, mdn, entry));
            gpsLogBatch.add(toGpsLogParams(vehicleId, mdn, entry));
            geofenceLogBatch.add(toGeofenceLogParams(vehicleId, mdn, entry));
        }

        jdbcTemplate.batchUpdate(POWER_LOG_INSERT_SQL, powerLogBatch);
        jdbcTemplate.batchUpdate(GPS_LOG_INSERT_SQL, gpsLogBatch);
        jdbcTemplate.batchUpdate(GEOFENCE_LOG_INSERT_SQL, geofenceLogBatch);

        log.info("차량 ID: {}에 대한 로그 데이터 배치 저장 완료", vehicleId);
    }

    private Object[] toPowerLogParams(Long vehicleId, Long mdn, LogDataEntry entry) {
        return new Object[]{
                vehicleId,
                mdn,
                "ON",
                entry.occurredTime(),
                entry.gpsStatus(),
                entry.latitude(),
                entry.longitude(),
                entry.totalTripMeter()
        };
    }

    private Object[] toGpsLogParams(Long vehicleId, Long mdn, LogDataEntry entry) {
        return new Object[]{
                vehicleId,
                mdn,
                entry.gpsStatus(),
                entry.latitude(),
                entry.longitude(),
                entry.angle(),
                entry.speed(),
                entry.totalTripMeter(),
                entry.batteryVoltage(),
                entry.occurredTime()
        };
    }

    private Object[] toGeofenceLogParams(Long vehicleId, Long mdn, LogDataEntry entry) {
        return new Object[]{
                vehicleId,
                mdn,
                entry.occurredTime(),
                entry.geofenceGroupId(),
                entry.geofenceId(),
                entry.eventVal(),
                entry.gpsStatus(),
                entry.latitude(),
                entry.longitude(),
                entry.angle()
        };
    }
}
```

SeoYoonP replied:
넵 저도 이 관리가 좋을것 같습니다. 반영해서 수정하겠습니다!

Jeongyounghyeon commented on src/main/java/org/thisway/log/dto/request/LogDataEntry.java:
```diff
@@ -0,0 +1,18 @@
+import java.time.LocalDateTime;
+
+public record LogDataEntry(
+        String gpsStatus, // "A", "V", "0" 중 하나
```
enum화 시켜도 좋을 것 같습니다!

SeoYoonP replied:
고려해본 후 다시 말씀드리겠습니다!

Jeongyounghyeon reviewed (commented):
연동 규격서를 이해하는 바로는 규격서의 규격을 지켜야한다고 생각했는데, 규격을 저희가 커스텀할 수 있는지에 대해 상의해봐야할 것 같습니다!

SeoYoonP commented:
규격서에대한 이해 부족으로 잘못된 구조를 구현했습니다. 구조를 다시 작성하여 PR리오픈 하겠습니다!

SeoYoonP committed:
refactor: #31 규격서에 따른 구조 변경으로 인한 기존 Log클래스들 삭제

- LogBatchRequest.java
- LogDataEntry.java
- LogDataSaveService.java
- LogDataSaveServiceTest.java

SeoYoonP committed:
feat: #31 GpsLog 데이터 전달 객체 정의

- GpsLogRequest
- GpsLogEntry

SeoYoonP committed:
feat: #31 PowerLog 데이터 전달 객체 정의

SeoYoonP committed:
feat: #31 GeofenceLog 데이터 전달 객체 정의

SeoYoonP committed:
feat: #31 로그 데이터 응답 객체 정의

SeoYoonP committed:
feat: #31 Emulator 엔티티 정의

SeoYoonP committed:
feat: #31 EmulatorRepository 정의

SeoYoonP committed:
feat: #31 에뮬레이터 ErrorCode 추가

SeoYoonP committed:
feat: #31 데이터 유형에 따른 LogController api 구현

SeoYoonP committed:
feat: #31 데이터 저장하는 LogService 구현

SeoYoonP committed:
feat: #31 LogService 단위 테스트 코드 작성

SeoYoonP committed:
Merge branch 'develop' into feat/31

Jeongyounghyeon commented on src/main/java/org/thisway/log/dto/request/gpsLog/GpsLogEntry.java:
해당 클래스만 `Request` 접미어를 제외한 이유가 있을까요?

SeoYoonP replied:
따로 Entry가 요청되는게 아니라 GpsLogRequest에 있는 cList에 포함되어 요청되는 개념이라 생각해서 Entry라는 이름으로 정의했습니다! cList <- [Entry]

Jeongyounghyeon replied:
class 네이밍에 요청 DTO라는 것을 파악할 수 있게 Request가 붙어도 좋을 것 같습니다!

SeoYoonP committed:
Merge branch 'develop' into feat/31

roomie00 commented on src/main/java/org/thisway/log/service/LogService.java:
```diff
@@ -0,0 +1,162 @@
+            });
+        }
+
+        String gpsLogSql = "INSERT INTO gps_log (vehicle_id, mdn, gps_status, latitude, longitude, anger, speed, total_trip_meter, battery_voltage, occurred_time) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
```
JDBC를 쓰더라도 SQL문은 repository 클래스를 별도로 만들어서 거기에 위치시키는게 좋을 것 같습니다.

SeoYoonP replied:
넵 말씀해주신 책임분리가 좋은 방향일 것같습니다. 현재 리팩토링 코드를 진행중에 있는데 따로 Repository를 만들어서 위치해두었습니다. 2차 mvp에서 반영할 계획입니다.
